File: .NET/VS2010TrainingKit/Labs/AspNetWebForms4/Source/Ex04-ViewState/end/VB/WebFormsSampleApp/Models/ShoppingCart.vb
' ----------------------------------------------------------------------------------
' Microsoft Developer & Platform Evangelism
' 
' Copyright (c) Microsoft Corporation. All rights reserved.
' 
' THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
' OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
' ----------------------------------------------------------------------------------
' The example companies, organizations, products, domain names,
' e-mail addresses, logos, people, places, and events depicted
' herein are fictitious.  No association with any real company,
' organization, product, domain name, email address, logo, person,
' places, or events is intended or should be inferred.
' ----------------------------------------------------------------------------------

Imports System.Collections.Generic
Imports System.Linq

Namespace WebFormsSampleApp.Models
	Public Class ShoppingCart
        Private _items As List(Of ShoppingCartItem)

        Public Sub New()
            _items = New List(Of ShoppingCartItem)()
        End Sub

        Public Property CartId() As String

        Public ReadOnly Property Items() As ShoppingCartItem()
            Get
                Return _items.ToArray()
            End Get
        End Property

        Public ReadOnly Property TotalItems() As Integer
            Get
                Return ( _
                    From i In _items _
                    Select i.Quantity).Sum()
            End Get
        End Property

        Public ReadOnly Property Subtotal() As Decimal
            Get
                If Items.Length = 0 Then
                    Return 0
                End If

                Dim result As Decimal = 0

                For Each item As ShoppingCartItem In Items
                    result += item.UnitPrice * item.Quantity
                Next item

                Return result
            End Get
        End Property

        Public Function AddItem(ByVal productId As Integer, ByVal productName As String, ByVal unitPrice As Decimal, ByVal quantity As Integer) As Decimal
            Dim item As ShoppingCartItem = Items.Where(Function(i) i.ProductId = productId).SingleOrDefault()

            If item IsNot Nothing Then
                item.Quantity += 1
            Else
                item = New ShoppingCartItem()
                item.ProductId = productId
                item.ProductName = productName
                item.UnitPrice = unitPrice
                item.Quantity = quantity
                _items.Add(item)
            End If

            Return Subtotal
        End Function
	End Class
End Namespace

File: .NET/VS2010TrainingKit/Labs/AspNetWebForms4/Source/Ex04-ViewState/end/C#/WebFormsSampleApp/Models/ShoppingCart.cs
// ----------------------------------------------------------------------------------
// Microsoft Developer & Platform Evangelism
// 
// Copyright (c) Microsoft Corporation. All rights reserved.
// 
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
// EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
// OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
// ----------------------------------------------------------------------------------
// The example companies, organizations, products, domain names,
// e-mail addresses, logos, people, places, and events depicted
// herein are fictitious.  No association with any real company,
// organization, product, domain name, email address, logo, person,
// places, or events is intended or should be inferred.
// ----------------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Web;

namespace WebFormsSampleApp.Models
{
    public class ShoppingCart
    {
        private List<ShoppingCartItem> items;

        public ShoppingCart()
        {
            items = new List<ShoppingCartItem>();
        }

        public string CartId { get; set; }

        public ShoppingCartItem[] Items
        {
            get
            {
                return items.ToArray();
            }
        }

        public int TotalItems
        {
            get
            {
                return (from i in items
                        select i.Quantity).Sum();
            }
        }

        public decimal Subtotal
        {
            get
            {
                if (Items.Length == 0) return 0;
                decimal subtotal = 0;
                foreach (ShoppingCartItem item in Items)
                {
                    subtotal += item.UnitPrice * item.Quantity;
                }
                return subtotal;
            }
        }

        public decimal AddItem(int productId, string productName, decimal unitPrice, int quantity)
        {
            ShoppingCartItem item = Items.Where(i => i.ProductId == productId).SingleOrDefault();
            if (item != null)
            {
                item.Quantity++;
            }
            else
            {
                item = new ShoppingCartItem();
                item.ProductId = productId;
                item.ProductName = productName;
                item.UnitPrice = unitPrice;
                item.Quantity = quantity;
                items.Add(item);
            }
            return Subtotal;
        }
    }
}
